src/app/clusters/group-key-mgmt-server/group-key-mgmt-server.cpp

```diff
@@ -0,0 +1,410 @@
+// Local functions
+//------------------------------------------------------------------------------
+
+static EmberAfStatus revokeKeys(uint16_t fabricIndex, uint16_t keySetIndex);
+static EmberAfStatus revokeKeys(uint16_t fabricIndex);
+
+static uint8_t nextEpoch(uint8_t epoch)
+{
+    if (epoch < GROUP_KEY_SET_SIZE - 1)
+    {
+        return epoch + static_cast<uint8_t>(1);
+    }
+    else
+    {
+        return 1;
+    }
+}
+
+static uint8_t previousEpoch(uint8_t epoch)
+{
+    if (epoch > 0)
+    {
+        return epoch - static_cast<uint8_t>(1);
+    }
+    else
+    {
+        return GROUP_KEY_SET_SIZE - 1;
+    }
+}
+
+static EmberAfStatus printKeys()
+{
+    emberAfGroupKeyManagementClusterPrint("GROUP KEY SETS(%u)", sGroupKeys.size());
+
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it;
+    uint16_t fabricIndex = 0;
+    uint16_t keySet      = 0;
+
+    for (it = sGroupKeys.begin(); it != sGroupKeys.end(); ++it)
+    {
+        fabricIndex = std::get<0>(it->first);
+        keySet      = std::get<1>(it->first);
+
+        emberAfGroupKeyManagementClusterPrint("  fabric:%u, set:%u", fabricIndex, keySet);
+        uint8_t index     = it->second->current < GROUP_KEY_SET_SIZE ? it->second->current : 0;
+        RootKey & current = it->second->epoch[index];
+        if (current.active)
+        {
+            emberAfGroupKeyManagementClusterPrint("    current: key(%u), start(%u), policy:%u", sizeof(current.value),
+                                                  sizeof(current.startTime), current.policy);
+
+            index              = previousEpoch(index);
+            RootKey & previous = it->second->epoch[index];
+            if (previous.active)
+            {
+                emberAfGroupKeyManagementClusterPrint("    previous: key(%u), start(%u), policy:%u", sizeof(previous.value),
+                                                      sizeof(previous.startTime), previous.policy);
+
+                index         = previousEpoch(index);
+                RootKey & old = it->second->epoch[index];
+                if (old.active)
+                {
+                    emberAfGroupKeyManagementClusterPrint("    old: key(%u), start(%u), policy:%u", sizeof(old.value),
+                                                          sizeof(old.startTime), old.policy);
+                }
+            }
+        }
+    }
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+static KeySet * getKeySet(uint16_t fabricIndex, uint16_t keySetIndex)
+{
+    std::tuple<uint16_t, uint16_t> index(fabricIndex, keySetIndex);
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it = sGroupKeys.find(index);
+    return it == sGroupKeys.end() ? nullptr : it->second;
+}
+
+static EmberAfStatus setKey(uint16_t fabricIndex, uint16_t keySetIndex, chip::ByteSpan & keyRoot, chip::ByteSpan & epochStartTime,
+                            uint8_t securityPolicy)
```

Comment:
`securityPolicy` is dropped to the floor and unused. The securityPolicy needs to be validated for being in the enum:

```
====== GroupKeySecurityPolicy

This field SHALL configure the security policy for an operational group key.

.Values of the _GroupKeySecurityPolicy_ Attribute
[cols="10,15,15,60",options="header"]
|===
| Value | Name        | Conformance | Description
| 0     | Standard    | M           | Message counter synchronization using <<MsgSyncCacheAndSync,cache-and-verify>>
| 1     | Low Latency | M           | Message counter synchronization using <<MsgSyncTrustFirst,trust-first>>
|===
```

The policy has to be stored for later validation at time of use of the keys.